modules/browsiRtdProvider.js

```diff
@@ -0,0 +1,253 @@
+}
+
+export function setData(data) {
+  _data = data;
+
+  if (typeof _dataReadyCallback === 'function') {
+    _dataReadyCallback(_data);
+    _dataReadyCallback = null;
+  }
+}
+
+/**
+ * wait for data from server
+ * call callback when data is ready
+ * @param {function} callback
+ */
+function waitForData(callback) {
+  if (_data) {
+    _dataReadyCallback = null;
+    callback(_data);
+  } else {
+    _dataReadyCallback = callback;
+  }
+}
+
+/**
+ * filter server data according to adUnits received
+ * call callback (onDone) when data is ready
+ * @param {adUnit[]} adUnits
+ * @param {function} onDone callback function
+ */
+function sendDataToModule(adUnits, onDone) {
+  try {
+    waitForData(_predictionsData => {
+      const _predictions = _predictionsData.p;
+      if (!_predictions || !Object.keys(_predictions).length) {
+        return onDone({});
+      }
+      const slots = getAllSlots();
+      if (!slots) {
+        return onDone({});
+      }
+      let dataToReturn = adUnits.reduce((rp, cau) => {
+        const adUnitCode = cau && cau.code;
+        if (!adUnitCode) { return rp }
+        const predictionData = _predictions[adUnitCode];
+        if (!predictionData) { return rp }
+
+        if (predictionData.p) {
+          if (!isIdMatchingAdUnit(adUnitCode, slots, predictionData.w)) {
+            return rp;
+          }
+          rp[adUnitCode] = getKVObject(predictionData.p, _predictionsData.kn);
+        }
+        return rp;
+      }, {});
+      return onDone(dataToReturn);
+    });
+  } catch (e) {
+    onDone({});
+  }
+}
+
+/**
+ * get all slots on page
+ * @return {Object[]} slot GoogleTag slots
+ */
+function getAllSlots() {
+  return utils.isGptPubadsDefined && window.googletag.pubads().getSlots();
+}
+/**
+ * get prediction and return valid object for key value set
+ * @param {number} p
+ * @param {string?} keyName
+ * @return {Object} key:value
+ */
+function getKVObject(p, keyName) {
+  const prValue = p < 0 ? 'NA' : (Math.floor(p * 10) / 10).toFixed(2);
+  let prObject = {};
+  prObject[((_moduleParams['keyName'] || keyName).toString())] = prValue.toString();
+  return prObject;
+}
+/**
+ * check if placement id matches one of given ad units
+ * @param {number} id placement id
+ * @param {Object[]} allSlots google slots on page
+ * @param {string[]} whitelist ad units
+ * @return {boolean}
+ */
+export function isIdMatchingAdUnit(id, allSlots, whitelist) {
+  if (!whitelist || !whitelist.length) {
+    return true;
+  }
+  const slot = allSlots.filter(s => s.getSlotElementId() === id);
+  const slotAdUnits = slot.map(s => s.getAdUnitPath());
+  return slotAdUnits.some(a => whitelist.indexOf(a) !== -1);
+}
+
+/**
+ * XMLHttpRequest to get data form browsi server
+ * @param {string} url server url with query params
+ */
+function getPredictionsFromServer(url) {
+  const xmlhttp = new XMLHttpRequest();
+  xmlhttp.onreadystatechange = function() {
```

Comment:
I think we may want to use the [Prebid Ajax](https://github.com/prebid/Prebid.js/blob/master/src/ajax.js) 

We do not really let modules / adapters handle their our requests usually. @jaiminpanchal27 does this seem right?

Reply:
https://github.com/prebid/Prebid.js/blob/master/modules/userId/unifiedIdSystem.js#L42 example of one of the user Id modules utilizing it

Reply:
Fixed